File: src/fwt/js/ComboPeer.js
//
// Copyright (c) 2009, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   1 Jun 09  Andy Frank  Creation
//

/**
 * ComboPeer.
 */
fan.fwt.ComboPeer = fan.sys.Obj.$extend(fan.fwt.WidgetPeer);
fan.fwt.ComboPeer.prototype.$ctor = function(self)
{
  this.m_items = fan.sys.List.make(fan.sys.Obj.$type);
}

// see init.js for CSS

fan.fwt.ComboPeer.prototype.font   = function(self) { return this.m_font; }
fan.fwt.ComboPeer.prototype.font$  = function(self, val) { this.m_font = val; }
fan.fwt.ComboPeer.prototype.m_font = null;

fan.fwt.ComboPeer.prototype.items   = function(self) { return this.m_items; }
fan.fwt.ComboPeer.prototype.items$  = function(self, val)
{
  this.m_items = val;
  this.needsRebuild = true;
}
fan.fwt.ComboPeer.prototype.m_items = null;

fan.fwt.ComboPeer.prototype.selectedIndex   = function(self) { return this.m_selectedIndex; }
fan.fwt.ComboPeer.prototype.selectedIndex$  = function(self, val)
{
  this.m_selectedIndex = val;
  if (this.elem != null && this.elem.firstChild != null)
    this.elem.firstChild.selectedIndex = val == null ? -1 : val;
}
fan.fwt.ComboPeer.prototype.m_selectedIndex = null;

fan.fwt.ComboPeer.prototype.text   = function(self) { return this.m_text; }
fan.fwt.ComboPeer.prototype.text$  = function(self, val) { this.m_text = val; }
fan.fwt.ComboPeer.prototype.m_text = "";

fan.fwt.ComboPeer.prototype.create = function(parentElem)
{
  // make sure we force rebuild
  this.needsRebuild = true;

  var select = document.createElement("select");
  select.className = "_fwt_Combo_";
  select.style.font = fan.fwt.WidgetPeer.fontToCss(this.m_font == null ? fan.fwt.DesktopPeer.$sysFont : this.m_font);

  var div = this.emptyDiv();
  div.appendChild(select);
  parentElem.appendChild(div);
  return div;
}

fan.fwt.ComboPeer.prototype.focus = function(self)
{
  if (this.elem == null) return
  var select = this.elem.firstChild;
  select.focus();
}

fan.fwt.ComboPeer.prototype.hasFocus = function(self)
{
  if (this.elem == null) return false;
  var select = this.elem.firstChild;
  return select === document.activeElement;
}

fan.fwt.ComboPeer.prototype.needsRebuild = true;
fan.fwt.ComboPeer.prototype.rebuild = function(self)
{
  // sync props
  var select = this.elem.firstChild;
  select.disabled = !this.m_enabled;

  // clear old items
  while (select.firstChild != null)
    select.removeChild(select.firstChild);

  // add new items
  for (var i=0; i<this.m_items.size(); i++)
  {
    var option = document.createElement("option");
    var text   = this.$itemText(self, this.m_items.get(i));
    option.appendChild(document.createTextNode(text));
    select.appendChild(option);
  }
}

fan.fwt.ComboPeer.prototype.sync = function(self)
{
  if (this.needsRebuild)
  {
    this.rebuild(self);
    this.needsRebuild = false;
  }

  // sync props
  var select = this.elem.firstChild;
  select.disabled = !this.m_enabled;

  // set selectedIndex to self to sync
  this.selectedIndex$(self, this.m_selectedIndex);

  select.onchange = function()
  {
    // sync changes back to widget
    self.selectedIndex$(select.selectedIndex);

    // fire onModify
    if (self.onModify().size() > 0)
    {
      var me = fan.fwt.Event.make();
      me.m_id = fan.fwt.EventId.m_modified;
      me.m_widget = self;
      var list = self.onModify().list();
      for (var i=0; i<list.size(); i++) list.get(i).call(me);
    }
  }

  fan.fwt.WidgetPeer.prototype.sync.call(this, self);
}

// Backdoor hook to override item text [returns Str]
fan.fwt.ComboPeer.prototype.$itemText = function(self, item)
{
  return fan.sys.ObjUtil.toStr(item);
}

File: src/fwt/java/ComboPeer.java
//
// Copyright (c) 2008, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   19 Jul 08  Brian Frank  Creation
//
package fan.fwt;

import fan.sys.*;
import fan.sys.List;
import org.eclipse.swt.*;
import org.eclipse.swt.graphics.Font;
import org.eclipse.swt.widgets.*;
import org.eclipse.swt.widgets.Event;
import org.eclipse.swt.widgets.Combo;
import org.eclipse.swt.widgets.Widget;
import org.eclipse.swt.events.*;

public class ComboPeer
  extends WidgetPeer
  implements ModifyListener
{

//////////////////////////////////////////////////////////////////////////
// Construction
//////////////////////////////////////////////////////////////////////////

  public static ComboPeer make(fan.fwt.Combo self)
    throws Exception
  {
    ComboPeer peer = new ComboPeer();
    ((fan.fwt.Widget)self).peer = peer;
    peer.self = self;
    return peer;
  }

  public Widget create(Widget parent)
  {
    fan.fwt.Combo self = (fan.fwt.Combo)this.self;

    int style = self.dropDown ? SWT.DROP_DOWN : SWT.SIMPLE;
    if (!self.editable) style |= SWT.READ_ONLY;
    Combo c = new Combo((Composite)parent, style);
    control = c;
    c.addListener(SWT.DefaultSelection, this);
    c.addModifyListener(this);
    c.setVisibleItemCount(20);
    return c;
  }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  // Str text := ""
  public String text(fan.fwt.Combo self) { return text.get(); }
  public void text(fan.fwt.Combo self, String v) { text.set(v); }
  public final Prop.StrProp text = new Prop.StrProp(this, "")
  {
    public String get(Widget w) { return ((Combo)w).getText(); }
    public void set(Widget w, String v) { ((Combo)w).setText(v); }
  };

  // Obj[] items := Obj[,]
  public List items(fan.fwt.Combo self) { return items.get(); }
  public void items(fan.fwt.Combo self, List v) { items.set(v); }
  public final Prop.ItemsProp items = new Prop.ItemsProp (this)
  {
    public void set(Widget w, String[] v) { ((Combo)w).setItems(v); }
  };

  // Font font := null
  public fan.gfx.Font font(fan.fwt.Combo self) { return font.get(); }
  public void font(fan.fwt.Combo self, fan.gfx.Font v) { font.set(v); }
  public final Prop.FontProp font = new Prop.FontProp(this)
  {
    public void set(Widget w, Font v) { ((Combo)w).setFont(v); }
  };

  // Int? selectedIndex := 0
  public Long selectedIndex(fan.fwt.Combo self) { return selectedIndex.get(); }
  public void selectedIndex(fan.fwt.Combo self, Long v) { selectedIndex.set(v != null ? v : -1); }
  public final Prop.IntProp selectedIndex = new Prop.IntProp(this, 0, true)
  {
    public int get(Widget w) { return ((Combo)w).getSelectionIndex(); }
    public void set(Widget w, int v) { if (v < 0) { ((Combo)w).deselectAll(); } else { ((Combo)w).select(v); } }
  };

//////////////////////////////////////////////////////////////////////////
// Eventing
//////////////////////////////////////////////////////////////////////////

  public void handleEvent(Event event)
  {
    fan.fwt.Combo self = (fan.fwt.Combo)this.self;
    if (event.type == SWT.DefaultSelection)
      self.onAction().fire(event(EventId.action));
    else
      super.handleEvent(event);
  }

  public void modifyText(ModifyEvent event)
  {
    fan.fwt.Combo self = (fan.fwt.Combo)this.self;
    self.onModify().fire(event(EventId.modified));
  }

}